File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    private boolean isvalid(int n, int m, int nrow, int ncol){
        return nrow >= 0 && ncol >= 0 && nrow < n && ncol < m;
    }
    private boolean isBound(int n, int m, int nrow, int ncol){
        return nrow == 0 || ncol == 0 || nrow == n - 1 || ncol == m - 1;
    }
    public void solve(char[][] board) {
        int n = board.length, m = board[0].length;
        boolean vis[][] = new boolean[n][m];
        int delrow[] = {1, 0, -1, 0};
        int delcol[] = {0, 1, 0, -1};
        for(int i = 0; i < n; i++)
        for(int j = 0; j < m; j++){
            if(board[i][j] == 'O' && !vis[i][j] && !isBound(n, m, i, j)){
                Queue<Pair<Integer, Integer>> q = new LinkedList<>();
                ArrayList<Pair<Integer, Integer>> temp = new ArrayList<>();
                Pair<Integer, Integer> comb = new Pair<>(i,j);
                q.add(comb);
                temp.add(comb);
                boolean flag = true;
                while(!q.isEmpty()){
                    int x = q.peek().getKey(), y = q.peek().getValue();
                    q.remove();
                    for(int k = 0; k < 4; k++){
                        int nrow = x + delrow[k];
                        int ncol = y + delcol[k];
                        if(isvalid(n,m,nrow,ncol) && !vis[nrow][ncol] && board[nrow][ncol] == 'O'){
                            vis[nrow][ncol] = true;
                            if(isBound(n, m, nrow, ncol))
                                flag = false;
                            comb = new Pair<>(nrow,ncol);
                            q.add(comb);
                            temp.add(comb);
                        }
                    }
                }
                if(flag){
                    for(Pair<Integer, Integer> it: temp)
                        board[it.getKey()][it.getValue()] = 'X';
                }
            }
        }
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
private:
    vector<int> nRow = {0, 1, 0, -1};
    vector<int> nCol = {1, 0, -1, 0};

    void dfs(int x, int y, vector<vector<char>> &board){
        board[x][y] = 'V';
        for(int i = 0; i < 4; i++){
            int nx = x + nRow[i];
            int ny = y + nCol[i];
            if(nx >= 0 && ny >= 0 && nx < board.size() && ny < board[0].size() && board[nx][ny] == 'O'){
                dfs(nx, ny, board);
            }
        }
    }
public:
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        for(int i = 0; i < n; i++){
            if(board[i][0] == 'O')
                dfs(i, 0, board);
            if(board[i][m - 1] == 'O')
                dfs(i, m - 1, board);
        }

        for(int i = 0; i < m; i++){
            if(board[0][i] == 'O')
                dfs(0, i, board);
            if(board[n - 1][i] == 'O')
                dfs(n - 1, i, board);
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(board[i][j] == 'V'){
                    board[i][j] = 'O';
                } else{
                    board[i][j] = 'X';
                }
            }
        }
    }
};